clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -547,34 +580,47 @@ private void updateGroupSubscription(Set<String> topics) {
         return groupAssignment;
     }
 
-    private void adjustAssignment(final Map<TopicPartition, String> ownedPartitions,
-                                  final Map<String, Assignment> assignments) {
-        boolean revocationsNeeded = false;
-        Set<TopicPartition> assignedPartitions = new HashSet<>();
+    /**
+     * Used by COOPERATIVE rebalance protocol only.
+     *
+     * Validate the user assignor returned assignments such that there are no partitions which are going to
+     * be assigned to a different consumer other than its current owner (i.e. who includes it in its owned
+     * partitions list encoded in subscription); also if there are any revoked partitions set the error code to let
+     * everyone retry.
+     */
+    private void validateAssignment(final Map<String, List<TopicPartition>> ownedPartitions,
```

Comment:
@ableegoldman @vvcephei @bbejeck This is the part of KIP-429 that would affect the assignor implementors the most: for any assignors claiming to support `COOPERATIVE` in its overridden `supportedProtocols` function (CooperativeStickyAssignor, StreamsAssignor), they need to respect the `owned-partitions` from member subscriptions; and here the ConsumerCoordinator would validate after reading it from the assignor's output.

If validation failed, it will set the fatal error code to all the member's assignment to tell them to stop; if any revocation needs to be triggered, it would also set the error code to let everyone rejoin immediately after completing the callbacks, which is just an optimization to reduce rebalance latency (so user's do not need to worry about setting themselves).